controllers/cluster_controller.go

```diff
@@ -123,6 +123,15 @@ func (r *ClusterReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, reterr e
 		return ctrl.Result{}, err
 	}
 
+	// Add finalizer first if not exist to avoid the race condition between init and delete
+	patched, err := patchHelper.PatchFinalizer(ctx, cluster, clusterv1.ClusterFinalizer)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if patched {
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
Have we tested that `Requeue` is actually needed here, I would expect the finalizer update to trigger an `Update` reconciliation on it's own.